msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractManagedIdentitySource.java

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.aad.msal4j;
+
+import com.nimbusds.oauth2.sdk.ParseException;
+import com.nimbusds.oauth2.sdk.SerializeException;
+import com.nimbusds.oauth2.sdk.http.HTTPRequest;
+import com.nimbusds.oauth2.sdk.http.HTTPResponse;
+import lombok.Getter;
+import lombok.Setter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.beans.Encoder;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+
+//base class for all sources that support managed identity
+abstract class AbstractManagedIdentitySource {
+
+    protected static final String TIMEOUT_ERROR = "[Managed Identity] Authentication unavailable. The request to the managed identity endpoint timed out.";
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractManagedIdentitySource.class);
+    private static final String MANAGED_IDENTITY_NO_RESPONSE_RECEIVED = "[Managed Identity] Authentication unavailable. No response received from the managed identity endpoint.";
+
+    protected final ManagedIdentityRequest managedIdentityRequest;
+    private ServiceBundle serviceBundle;
+    private ManagedIdentitySourceType managedIdentitySourceType;
+
+    @Getter
+    @Setter
+    private boolean isUserAssignedManagedIdentity;
+    @Getter
+    @Setter
+    private String managedIdentityUserAssignedClientId;
+    @Getter
+    @Setter
+    private String managedIdentityUserAssignedResourceId;
+
+    public AbstractManagedIdentitySource(MsalRequest msalRequest, ServiceBundle serviceBundle,
+                                         ManagedIdentitySourceType sourceType) {
+        this.managedIdentityRequest = (ManagedIdentityRequest) msalRequest;
+        this.managedIdentitySourceType = sourceType;
+        this.serviceBundle = serviceBundle;
+    }
+
+    public ManagedIdentityResponse getManagedIdentityResponse(
+            ManagedIdentityParameters parameters) {
+
+        createManagedIdentityRequest(parameters.resource);
+        IHttpResponse response;
+
+        try {
+            HttpRequest httpRequest = new HttpRequest(HttpMethod.GET, managedIdentityRequest.computeURI().toString(), managedIdentityRequest.headers);
+            response = HttpHelper.executeHttpRequest(httpRequest, managedIdentityRequest.requestContext(), serviceBundle);
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+
+        return handleResponse(parameters, response);
+    }
+
+    public ManagedIdentityResponse handleResponse(
+            ManagedIdentityParameters parameters,
+            IHttpResponse response) {
+
+        String message;
+
+        try {
+            if (response.statusCode() == HttpURLConnection.HTTP_OK) {
```

Comment:
Can this ever be a different code and still return a good response, like HTTP 304?

Reply:
I don't think so. There are a set status codes that an endpoint returns. We did not observe this behavior in testing. Also, when we implemented this in MSAL.Net most of the logic was copied from Azure.Identity along with this where we check 200.